include/exec/repeat_effect_until.hpp

```diff
@@ -0,0 +1,309 @@
+      : __source_(std::move(__source)), __predicate_(std::move(__predicate)),
+        __receiver_(std::move(__r)) {
+    __source_op_.__construct_with([&] {
+      return stdexec::connect(std::move(__source_), __receiver_t{this});
+    });
+  }
+
+  ~__t() {
+    if (__is_source_op_constructed_) {
+      __source_op_.__destruct();
+      __is_source_op_constructed_ = false;
+    }
+  }
+
+  friend void tag_invoke(stdexec::start_t, __t &__self) noexcept {
+    stdexec::start(__self.__source_op_.__get());
+  }
+};
+
+template <typename _SourceId, typename _Predicate> struct __sender {
+  using _Source = stdexec::__t<_SourceId>;
+
+  template <class _Receiver>
+  using __op_t =
+      stdexec::__t<__operation<_SourceId, _Predicate,
+                               stdexec::__id<std::remove_cvref_t<_Receiver>>>>;
+
+  template <class _Receiver>
+  using __receiver_t =
+      stdexec::__t<__receiver<_SourceId, _Predicate,
+                              stdexec::__id<std::remove_cvref_t<_Receiver>>>>;
+
+  struct __t {
+    using __id = __sender;
+    using is_sender = void;
+    [[no_unique_address]] _Source __source_;
+    [[no_unique_address]] _Predicate __predicate_;
+
+    template <class _Self, class _Env>
+    using __completion_signatures = //
+        __make_completion_signatures<
+            __copy_cvref_t<_Self, _Source>, _Env,
+            __with_error_invoke_t<set_value_t, _Predicate,
+                                  __copy_cvref_t<_Self, _Source>, _Env>>;
+
+    template <__decays_to<__t> _Self, class _Env>
+    friend auto tag_invoke(get_completion_signatures_t, _Self &&, _Env)
+        -> __completion_signatures<_Self, _Env>
+    requires true;
+
+    template <typename _Source2, typename _Predicate2>
+    explicit __t(_Source2 &&__source, _Predicate2 &&__predicate) noexcept(
+        std::is_nothrow_constructible_v<_Source, _Source2>
+            &&std::is_nothrow_constructible_v<_Predicate, _Predicate2>)
+        : __source_(std::move(__source)),
+          __predicate_(std ::move(__predicate)) {}
+
+    template <typename _Sender, receiver _Receiver>
+    requires same_as<std::remove_cvref_t<_Sender>, __t> &&
+        sender_to<_Source, __receiver_t<_Receiver>>
+    friend __op_t<_Receiver> tag_invoke(connect_t, _Sender &&__s,
+                                        _Receiver &&__r) {
+      return {std::move(__s).__source_, std::move(__s).__predicate_,
+              std::move(__r)};
+    }
+
+    friend env_of_t<const _Source &> tag_invoke(get_env_t,
+                                                const __t &__self) noexcept {
+      return get_env(__self.__source_);
+    }
+  };
+};
+
+template <typename _Source, typename _Predicate>
+using __sender_p = __t<__sender<stdexec::__id<std::remove_cvref_t<_Source>>,
+                                std::decay_t<_Predicate>>>;
+
+inline const struct repeat_effect_until_t {
+  template <sender _Source, __movable_value _Predicate>
```

Comment:
Shouldn't this require that `_Predicate` is a predicate instead of `__movable_value` ?

Reply:
A predicate is a function that returns `bool` when passed an argument. Here we don't know the type of the argument yet, so we can't test that `_Predicate` is actually a predicate. We should test that in the `get_completion_signatures` customization.

Reply:
Thanks. In set_value, I checked if the value that the sender generated can be converted to bool.